python/lsst/daf/butler/tests/utils.py

```diff
@@ -55,3 +67,184 @@ def assertAstropyTablesEqual(self, tables, expectedTables):
+
+        # New datasets will be added to run and tag, but we will only look in
+        # tag when looking up datasets.
+        run = "ingest/run"
+        tag = "ingest"
+        self.butler = Butler(butlerConfigFile, run=run, collections=[tag], tags=[tag])
+
+        # Create and register a DatasetType
+        self.datasetType = self.addDatasetType(dimensions=("instrument", "visit"),
+                                               datasetTypeName="test_metric_comp",
+                                               storageClass="StructuredCompositeReadComp")
+
+        # Add needed Dimensions
+        self.addDimensionData("instrument", {"name": "DummyCamComp"})
+        self.addDimensionData("physical_filter", {"instrument": "DummyCamComp",
+                                                  "name": "d-r",
+                                                  "band": "R"})
+        self.addDimensionData("visit_system", {"instrument": "DummyCamComp",
+                                               "id": 1,
+                                               "name": "default"})
+        visitStart = astropy.time.Time("2020-01-01 08:00:00.123456789", scale="tai")
+        visitEnd = astropy.time.Time("2020-01-01 08:00:36.66", scale="tai")
+        self.addDimensionData("visit", dict(instrument="DummyCamComp",
+                                            id=423,
+                                            name="fourtwentythree",
+                                            physical_filter="d-r",
+                                            visit_system=1,
+                                            datetimeBegin=visitStart,
+                                            datetimeEnd=visitEnd))
+        self.addDimensionData("visit", dict(instrument="DummyCamComp",
+                                            id=424,
+                                            name="fourtwentyfour",
+                                            physical_filter="d-r",
+                                            visit_system=1))
+
+        self.addDataset({"instrument": "DummyCamComp", "visit": 423})
+        self.addDataset({"instrument": "DummyCamComp", "visit": 424})
+
+    def addDataset(self, dataId=None, run=None, datasetType=None):
+        """Create a new example metric and add it to the named run with the
+        given dataId.
+
+        Parameters
+        ----------
+        dataId : `dict`
+            The dataId for the new metric.
+        run : `str`, optional
+            The name of the run to create and add a dataset to. If `None`, the
+            dataset will be added to the root butler.
+        datasetType : ``DatasetType``, optional
+            The dataset type of the added dataset. If `None`, will use the
+            default dataset type.
+        """
+        if not datasetType:
+            datasetType = self.datasetType
+        ref = DatasetRef(datasetType, dataId, id=None)
+        metric = self._makeExampleMetrics()
+        if run:
```

Comment:
It fails when I call this and the run doesn't exist yet.

Reply:
It would be clearer then if you had a line here that created the run collection rather than without commentary relying on Butler to create it behind the scenes.

Reply:
maybe it's weird if/because I'm thinking about it wrong though...

Reply:
Without a comment here it all seems superfluous. Whereas:

```python
if run:
    self.butler.registry.registerCollection(self.run, type=CollectionType.RUN)
```
is explicit about what you need. @TallJimbo do you prefer the "create a new butler for each run" paradigm? Maybe that's better for the general user. If that is the case I'd like a comment to explain why you are having to create a new butler.

Reply:
I see. That makes sense, works.

Reply:
Outside of tests, users are never going to create new runs or even call `put` themselves (it'll all be operations like ingest or `pipetask` doing that).  So I think we can go with whatever maximizes test code readability.  I don't have a super strong preference; the "new butler for each run" pattern is maybe a bit more familiar to those coming from Gen2, but I agree that it's less explicit.